src/main/java/com/atlassian/db/replica/api/DualConnection.java

```diff
@@ -75,8 +107,13 @@ public CallableStatement prepareCall(String sql) throws SQLException {
     @Override
     public String nativeSQL(String sql) throws SQLException {
         checkClosed();
-        return connectionProvider.getWriteConnection(new RouteDecisionBuilder(Reason.RW_API_CALL).sql(sql)).nativeSQL(
-            sql);
+        if (experiments.isEnabled("com.atlassian.db.replica.experiment.nativesql.on.replica")) {
```

Comment:
This is a new behavior the consumer wants to rely on. It's de facto API. So it's better make it explicit. And if a magic string is API, it's better to turn it to a method. Like `Experiments.shouldRunNativeSqlOnReplica()` with a default impl of `false`. Then after it's deemed stable, the SPI can remain as is without changing the default. The consumer can just hardcode it to `true` on their side. We can add Javadoc saying that `true` was well tested. Eventually, in a major release we can change the default to `true` or remove the toggle. But that's not necessary neither in the short term nor in the long term.

- [ ] reconsider magic string contract

Reply:
> This is a new behavior the consumer wants to rely on.

Experiments by definition are short-lived. We want to get rid of them as fast as we gain the confidence it works well. The only contract I would give for them is to be there for one release. It may and should be cleaned up with the next release. By changing the contracts from strings to methods, we can't remove the flags until the major release. The API will only grow.

Keeping the toggles that long would change the project into a maintenance nightmare with tens of flags that may manipulate the state.

I imagine this API like a zero-downtime upgrade feature. You can bump the library safely and control all the changes (revert if needed), but you can do it only from one version to the next version. I would not commit to providing an upgrade path for all features up to a major release.

Reply:
There's no "short-lived" API in SemVer. If the toggles would actually become a maintenance nightmare it just means we need a major release more often.

Reply:
`Experiment` is API. The behaviour for the changes behind the flags is not. Unfortunately, I have no better idea not to promise it will always work (especially there's no such use case. You either need it for rollouts or you don't need it at all. We don't need the flag after it's stable).

Reply:
Then any experiment can be broken at any lib bump. If that's ok by the consumers, then we can proceed

Reply:
> Then any experiment can be broken at any lib bump. If that's ok by the consumers, then we can proceed

Yes. Any lib bump can introduce errors. It's the standard way libs operate. `Experiment` is for safe bumps. Customers can also have other (better) ways for updates (OSGI for example), unfortunately, we don't use it.

Reply:
The consumer will make a typo in the magic string or will use a string that's no longer supported. It will not be detected at bump time or at compile time. It *might* get caught in runtime, but it can also remain hidden, because it won't throw exceptions.

Experiments will be used (for a short timer or not). If they don't affect public API (circumventing SemVer), then we'd need to maintain a separate list of supported experiments and when they were introduced and when they were dropped. `EXPERIMENT_LOG.md`?

I don't see why these new problems would be smaller than just releasing a major bump.

Reply:
Also see [Martin Fowler's `FeatureDecisions` example](https://www.martinfowler.com/articles/feature-toggles.html#De-couplingDecisionPointsFromDecisionLogic). It's much more powerful (and discoverable) than a `Map<String, boolean>`.

Reply:
For consumer mistakes, and referring to the nature of [hypothesis-driven-development](https://www.thoughtworks.com/insights/articles/how-implement-hypothesis-driven-development) I'd consider adding automatic rollback of experiment/hypothesis, supported by circuit breaker - too many exceptions ? rolled it back automatically, or give a customer a way to skip automatic rollback, but not by default.